### PR TITLE
CXP-1239: Catalog Completeness criterion Front

### PR DESCRIPTION
**Pull request by RiffVentura**
Added a system criterion: completeness

### CODE REVIEW

File: components/catalogs/front/src/components/ProductSelection/criteria/CompletenessCriterion/CompletenessCriterion.integration.tsx [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+        scope: 'ecommerce',
+    });
+});
+
+function changeOperatorValueTo(operator: string) {
+    changeSelectValueTo('operator', operator);
+}
+
+function changeChannelValueTo(operator: string) {
+    changeSelectValueTo('scope', operator);
```
Comment (masacc) [inline components/catalogs/front/src/components/ProductSelection/criteria/CompletenessCriterion/CompletenessCriterion.integration.tsx:330]:
```suggestion
function changeChannelValueTo(channel: string) {
    changeSelectValueTo('scope', channel);
```

File: components/catalogs/front/src/components/ProductSelection/criteria/CompletenessCriterion/CompletenessCriterion.integration.tsx [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+}
+
+function changeLocaleValueTo(operator: string) {
+    changeSelectValueTo('locale', operator);
```
Comment (masacc) [inline components/catalogs/front/src/components/ProductSelection/criteria/CompletenessCriterion/CompletenessCriterion.integration.tsx:334]:
```suggestion
function changeLocaleValueTo(locale: string) {
    changeSelectValueTo('locale', locale);
```

File: components/catalogs/front/src/components/ProductSelection/hooks/useInfiniteChannels.unit.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    });
+
+    expect(result.current).toMatchObject({
+        isLoading: false,
+        isError: false,
+        data: [channelPrint],
+        error: null,
+        hasNextPage: false,
+        fetchNextPage: expect.any(Function),
+    });
```
Comment (masacc) [inline components/catalogs/front/src/components/ProductSelection/hooks/useInfiniteChannels.unit.ts:84]: duplicated code

File: components/catalogs/front/src/components/ProductSelection/hooks/useInfiniteChannels.unit.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
```
Comment (masacc) [inline components/catalogs/front/src/components/ProductSelection/hooks/useInfiniteChannels.unit.ts:97]: is it really useful to test this in this use case?

**Review by masacc — APPROVED** [review]
Really nice job on this PR, this is clean and clear :clap:

File: components/catalogs/front/src/components/ProductSelection/criteria/CompletenessCriterion/CompletenessCriterion.integration.tsx [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+    expect(onChange).toHaveBeenCalledWith({
+        field: 'completeness',
+        operator: Operator.LOWER_THAN,
+        value: 25,
+        locale: null,
+        scope: 'ecommerce',
+    });
+});
+
+function changeOperatorValueTo(operator: string) {
```
Comment (tseho) [inline components/catalogs/front/src/components/ProductSelection/criteria/CompletenessCriterion/CompletenessCriterion.integration.tsx:325]: Small suggestion, can you move the helpers to the top of the file ?
It's really unusual to find them at the bottom.
Also, I don't think we use the `function` syntax anywhere.
You should do `const changeOperatorTo = (value: string) => changeSelectValueTo('operator', value);` IMO.

File: components/catalogs/front/src/components/ProductSelection/criteria/CompletenessCriterion/factory.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import {CompletenessCriterionState} from './types';
+import {Operator} from '../../models/Operator';
+
+export default (state?: Partial<CompletenessCriterionState>): CompletenessCriterionState => ({
+    field: 'completeness',
+    operator: state?.operator ?? Operator.EQUALS,
+    value: state?.value ?? 100,
+    locale: state?.locale ?? null,
+    scope: state?.scope ?? null,
```
Comment (tseho) [inline components/catalogs/front/src/components/ProductSelection/criteria/CompletenessCriterion/factory.ts:9]: I don't care if you change how it's written, but can you also rewrite the other two so they stay consistent ?

File: components/catalogs/front/src/components/ProductSelection/criteria/FamilyCriterion/FamilyOperatorInput.tsx [hunk 7]
```diff
@@ -28,27 +29,35 @@ const FamilyOperatorInput: FC<Props> = ({state, onChange}) => {
...
+                </SelectInput.Option>
+                <SelectInput.Option value={Operator.NOT_IN_LIST} title={translateOperator(Operator.NOT_IN_LIST)}>
+                    {translateOperator(Operator.NOT_IN_LIST)}
+                </SelectInput.Option>
+            </SelectInput>
+            {error !== undefined && (
+                <Helper inline level='error'>
+                    {error}
+                </Helper>
+            )}
```
Comment (tseho) [inline components/catalogs/front/src/components/ProductSelection/criteria/FamilyCriterion/FamilyOperatorInput.tsx:59]: I've noticed that you display the errors at the correct place for the completeness but it's at the wrong place for family & status (it's my fault).
Can you move them, the same way you did in completeness ?

File: components/catalogs/front/src/components/ProductSelection/hooks/useInfiniteChannels.unit.ts (hunk 4 above)
Comment (tseho) [inline components/catalogs/front/src/components/ProductSelection/hooks/useInfiniteChannels.unit.ts:97]:
```suggestion
```

Same comment as masacc, those 3 tests are useless.

**Review by tseho — APPROVED** [review]

**Comment by tseho** [comment]
I agree with masacc, great job on this PR, it was really easy to read :+1:

File: components/catalogs/front/src/components/ProductSelection/hooks/useScopedLocales.ts [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import {Locale} from '../models/Locale';
+import {useInfiniteChannels} from './useInfiniteChannels';
+
+export const useScopedLocales = (channelCode: string | null): Locale[] => {
+    const {data: channels, isLoading, isError, error} = useInfiniteChannels({code: channelCode});
```
Comment (masacc) [inline components/catalogs/front/src/components/ProductSelection/hooks/useScopedLocales.ts:5]: replace this hook by one that fetch only one channel (create the new endpoint on the back)

File: components/catalogs/front/src/components/ProductSelection/criteria/ErrorHelpers.tsx [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+                return null;
+            }
+
+            return (
+                <Helper key={key} level='error'>
+                    {errors[key]}
+                </Helper>
+            );
+        })
+        .filter(element => null !== element);
```
Comment (tseho) [inline components/catalogs/front/src/components/ProductSelection/criteria/ErrorHelpers.tsx:22]: why don't you filter first ?

File: components/catalogs/front/src/components/ProductSelection/criteria/ErrorHelpers.tsx [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+    if (0 === errorHelpers.length) {
+        return <></>;
```
Comment (tseho) [inline components/catalogs/front/src/components/ProductSelection/criteria/ErrorHelpers.tsx:25]:
```suggestion
        return null;
```

File: components/catalogs/front/src/components/ProductSelection/criteria/FamilyCriterion/FamilyOperatorInput.tsx [hunk 11]
```diff
@@ -28,27 +29,30 @@ const FamilyOperatorInput: FC<Props> = ({state, onChange}) => {
...
+                    {translateOperator(Operator.IS_NOT_EMPTY)}
+                </SelectInput.Option>
+                <SelectInput.Option value={Operator.IN_LIST} title={translateOperator(Operator.IN_LIST)}>
+                    {translateOperator(Operator.IN_LIST)}
+                </SelectInput.Option>
+                <SelectInput.Option value={Operator.NOT_IN_LIST} title={translateOperator(Operator.NOT_IN_LIST)}>
+                    {translateOperator(Operator.NOT_IN_LIST)}
+                </SelectInput.Option>
+            </SelectInput>
+        </>
```
Comment (tseho) [inline components/catalogs/front/src/components/ProductSelection/criteria/FamilyCriterion/FamilyOperatorInput.tsx:55]: You can revert this change, the error is not displayed there anymore.

File: components/catalogs/front/src/components/ProductSelection/criteria/FamilyCriterion/FamilySelectInput.tsx [hunk 12]
```diff
@@ -9,33 +9,37 @@ import {useTranslate} from '@akeneo-pim-community/shared';
...
+                invalid={isInvalid}
+                data-testid='value'
+            >
+                {families.map(family => (
+                    <MultiSelectInput.Option key={family.code} title={family.label} value={family.code}>
+                        {family.label}
+                    </MultiSelectInput.Option>
+                ))}
+            </MultiSelectInput>
+        </>
```
Comment (tseho) [inline components/catalogs/front/src/components/ProductSelection/criteria/FamilyCriterion/FamilySelectInput.tsx:42]: Same, can be reverted.

File: components/catalogs/front/src/components/ProductSelection/criteria/FamilyCriterion/FamilyOperatorInput.tsx [hunk 13]
```diff
@@ -28,27 +29,30 @@ const FamilyOperatorInput: FC<Props> = ({state, onChange}) => {
...
-                {translateOperator(Operator.IS_NOT_EMPTY)}
-            </SelectInput.Option>
-            <SelectInput.Option value={Operator.IN_LIST} title={translateOperator(Operator.IN_LIST)}>
-                {translateOperator(Operator.IN_LIST)}
-            </SelectInput.Option>
-            <SelectInput.Option value={Operator.NOT_IN_LIST} title={translateOperator(Operator.NOT_IN_LIST)}>
-                {translateOperator(Operator.NOT_IN_LIST)}
-            </SelectInput.Option>
-        </SelectInput>
+        <>
```
Comment (tseho) [inline components/catalogs/front/src/components/ProductSelection/criteria/FamilyCriterion/FamilyOperatorInput.tsx:32]:
```suggestion
```
You can revert this change, the error is not displayed there anymore, no need for `<></>`.

File: components/catalogs/front/src/components/ProductSelection/criteria/FamilyCriterion/FamilySelectInput.tsx [hunk 14]
```diff
@@ -9,33 +9,37 @@ import {useTranslate} from '@akeneo-pim-community/shared';
...
-            onSearchChange={setSearch}
-            data-testid='value'
-        >
-            {families.map(family => (
-                <MultiSelectInput.Option key={family.code} title={family.label} value={family.code}>
-                    {family.label}
-                </MultiSelectInput.Option>
-            ))}
-        </MultiSelectInput>
+        <>
```
Comment (tseho) [inline components/catalogs/front/src/components/ProductSelection/criteria/FamilyCriterion/FamilySelectInput.tsx:23]:
```suggestion
```
Same, can be removed.

File: components/catalogs/front/src/components/ProductSelection/criteria/CompletenessCriterion/CompletenessOperatorInput.tsx [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+            </SelectInput.Option>
+            <SelectInput.Option value={Operator.NOT_EQUAL} title={translateOperator(Operator.NOT_EQUAL)}>
+                {translateOperator(Operator.NOT_EQUAL)}
+            </SelectInput.Option>
+            <SelectInput.Option value={Operator.LOWER_THAN} title={translateOperator(Operator.LOWER_THAN)}>
+                {translateOperator(Operator.LOWER_THAN)}
+            </SelectInput.Option>
+            <SelectInput.Option value={Operator.GREATER_THAN} title={translateOperator(Operator.GREATER_THAN)}>
+                {translateOperator(Operator.GREATER_THAN)}
+            </SelectInput.Option>
```
Comment (RiffVentura) [inline components/catalogs/front/src/components/ProductSelection/criteria/CompletenessCriterion/CompletenessOperatorInput.tsx:47]:
```suggestion
            {[
                Operator.EQUALS,
                Operator.NOT_EQUAL,
                Operator.LOWER_THAN,
                Operator.GREATER_THAN,
            ].map(operator => (
                <SelectInput.Option key={operator} value={operator} title={translateOperator(operator)}>
                    {translateOperator(operator)}
                </SelectInput.Option>
            ))}
```
:thinking:

**Review by tseho — APPROVED** [review]
The PR was already of great quality, and @masacc, you did a really great job on the refactoring to make it perfect.

:clap:

File: components/catalogs/front/src/components/ProductSelection/criteria/CompletenessCriterion/CompletenessOperatorInput.tsx (hunk 15 above)
Comment (tseho) [reply components/catalogs/front/src/components/ProductSelection/criteria/CompletenessCriterion/CompletenessOperatorInput.tsx:47]: Yeah, I wondered the same, IMO, it's not required to be changed in this PR because we should update existing criteria to be consistent.
It's a cleanup task we can add in our backlog.

File: components/catalogs/front/src/components/ProductSelection/criteria/ErrorHelpers.tsx (hunk 9 above)
Comment (masacc) [reply components/catalogs/front/src/components/ProductSelection/criteria/ErrorHelpers.tsx:22]: The logic is a bit awkward: you need to transform all values of an object into something else (components here) and I think you can't filter on object properties. Do you agree?

Comment (tseho) [reply components/catalogs/front/src/components/ProductSelection/criteria/ErrorHelpers.tsx:22]: I think that 

```
        .filter(key => errors[key] !== undefined);
        .map(key => {
            return (
                <Helper key={key} level='error'>
                    {errors[key]}
                </Helper>
            );
        })
```
should work

**Review by guillaumefer-akeneo — APPROVED** [review]
